### PR TITLE
Introduced SmartHighlight Mode

### PR DESCRIPTION
**Pull request by Tiax**
- Introduced "SmartHighlight Mode" into settings dialog.
- SmartHighlighting now highlights based on the set mode (whole word, any selection, same as find dialog).

See issue #426 for additional info

### CODE REVIEW

File: PowerEditor/src/Parameters.cpp [hunk 1]
```diff
@@ -4020,6 +4020,31 @@ void NppParameters::feedGUIParameters(TiXmlNode *node)
 			}
 		}
 
+		else if (!lstrcmp(nm, TEXT("SmartHighLightMode")))
```
Comment (MAPJe71) [inline PowerEditor/src/Parameters.cpp]: Please do not use the `lstr`-functions, use the `_tcs` functions instead.

Comment (Tiax) [inline PowerEditor/src/Parameters.cpp]: Seems like a bad habit to change it mid-way: I would say either refactor the entire file or stick to the methods already used there.

**Comment by cmeriaux** [comment]
:+1 This new feature is very useful. Thanks you very much

Comments: when I see the quantity of code you must write for adding a single parameter in the GUI, it makes me feel that GUI is painful to write. We should trash GUI parameters and do everything with Configuration Files (Like SublimeText)

**Comment by dail8859** [comment]
@cmeriaux Yes it is unfortunate that in several cases the "supporting" code (GUI, parameters, etc) out-weigh the actual feature.

File: PowerEditor/src/Parameters.cpp [hunk 2]
```diff
@@ -4020,6 +4020,31 @@ void NppParameters::feedGUIParameters(TiXmlNode *node)
...
+		{
+			TiXmlNode *n = childNode->FirstChild();
+
+			if (n)
+			{
+				const TCHAR* val = n->Value();
+				if (val)
+				{
+					if (!lstrcmp(val, TEXT("1")))
```
Comment (dail8859) [inline PowerEditor/src/Parameters.cpp]: Is this if/else block necessary (and subsequent checks when writing the parameter)? Can the pure enum value be stored instead?

If you take a look at the "_autoCAction" attribute in this file it uses its enum value.

Comment (Tiax) [inline PowerEditor/src/Parameters.cpp]: True, we could probably do a static_cast; I personally don't think it makes it any more readable or more performant, though.

**Pull request closed without merging**